app/controllers/IndexCommon.java

```diff
@@ -0,0 +1,341 @@
+      } else {
+        return created("Created " + resource.getId());
+      }
+    }
+  }
+
+  protected Result addListProcessingReport(ModelCommon resource, ProcessingReport processingReport) {
+    if (!processingReport.isSuccess()) {
+      ListProcessingReport listProcessingReport = new ListProcessingReport();
+      try {
+        listProcessingReport.mergeWith(processingReport);
+      } catch (ProcessingException e) {
+        Logger.warn("Failed to create list processing report", e);
+      }
+      if (request().accepts("text/html")) {
+        Map<String, Object> scope = new HashMap<>();
+        scope.put("report", OBJECT_MAPPER.convertValue(listProcessingReport.asJson(), ArrayList.class));
+        scope.put("type", resource.getType());
+        return badRequest(render("Upsert failed", "ProcessingReport/list.mustache", scope));
+      } else {
+        return badRequest(listProcessingReport.asJson());
+      }
+    }
+    return null;
+  }
+
+  protected abstract Result upsertItems() throws IOException;
+
+  protected Result upsertItems(final String... aForbiddenTypes) throws IOException {
+
+    // Extract resources
+    List<ModelCommon> resources = new ArrayList<>();
+    for (JsonNode jsonNode : getJsonFromRequest()) {
+      Resource resource = new Resource(jsonNode);
+      resource.put(JsonLdConstants.CONTEXT, mConf.getString("jsonld.context"));
+      resources.add(resource);
+    }
+
+    // Validate
+    ListProcessingReport listProcessingReport = new ListProcessingReport();
+    for (ModelCommon resource : resources) {
+      // Person create /update only through UserIndex, which is restricted to admin
+      for (String forbiddenType : aForbiddenTypes) {
+        if (forbiddenType.equals(resource.getType())) {
+          return forbidden(String.format("Upsert %s forbidden.", forbiddenType));
+        }
+      }
+
+      // Stage and validate each resource
+      if (!stageAndValidate(listProcessingReport, resource)) {
+        return badRequest();
+      }
+    }
+
+    if (!listProcessingReport.isSuccess()) {
+      return badRequest(listProcessingReport.asJson());
+    }
+    mBaseRepository.addItems(resources, getMetadata());
+    return ok("Added resources");
+  }
+
+  protected boolean stageAndValidate(ListProcessingReport listProcessingReport, ModelCommon resource) throws IOException {
+    try {
+      ModelCommon staged = mBaseRepository.stage(resource);
+      ProcessingReport processingMessages = staged.validate(mTypes.getSchema(staged.getClass()));
+      if (!processingMessages.isSuccess()) {
+        Logger.debug(processingMessages.toString());
+        Logger.debug(staged.toString());
+      }
+      listProcessingReport.mergeWith(processingMessages);
+    } catch (ProcessingException e) {
+      Logger.error("Could not process validation report", e);
+      return false;
+    }
+    return true;
+  }
+
+  public Result read(String id, String version, String extension) throws IOException {
```

Comment:
How does this relate to `ResourceIndex.read`?

Reply:
Must be `abstract` here. Next commit.